### DISCUSSION TITLE
FFT : cufft backend

### DISCUSSION THREAD

**Discussion by pfeatherstone** [davisking/dlib#2756] [Ideas]
I suggest maybe adding a cuFFT backend implementation of `dlib::fft`. Maybe we give it another name like `dlib::cu::fft` so that applications can use both CPU and GPU. This won't be useful for small FFTs but sizes >= 1024x1024 this will definitely help. I did a quick test with FFT size 32x1024x1024.  With MKL it took around 400ms (single threaded). With cuFFT it took around 3ms. So this is a win.

**Comment by pfeatherstone** [comment]


**Reply by arrufat** [reply]
There is already `dlib::cuda::set_device(int dev);` :)

**Reply by pfeatherstone** [reply]
Ah sorry. I'll delete this thread.

**Comment by davisking** [comment]
Yeah that would be neat.  That timing you did includes the transfer time to and from the GPU? Normally you want to organize the code to hide those transfer times.

**Reply by pfeatherstone** [reply]
Ok, you are right. Here are the benchmarks using the following.
I'm using Pytorch as the API is easy to use when switching between devices and I'm assuming it's using MKL for CPU.

```
ntests      = 1000
v           = torch.randn(32*1024*1024, dtype=torch.complex64).to(device_host)

start = time.time()
for i in range(ntests):
    V = torch.fft.fft(v.to(device_comp)).to(device_host)
stop = time.time()
print("Average FFT time {}ms".format((stop - start) * 1e3 / ntests))
```

Using:
```
device_host = 'cuda:0'
device_comp = 'cuda:0'
```
I get **3.334** ms per FFT.

Using:
 ```
device_host = 'cpu'
device_comp = 'cuda:0'
```
I get **186.005** ms per FFT

Using:
```
device_host = 'cpu'
device_comp = 'cpu'
torch.set_num_threads(1)
```
I get **526.02** ms per FFT

**Reply by pfeatherstone** [reply]
When using MKL directly, it was more like 430ms but that's close enough. So yes, there is a massive penalty when including transfer between CPU and GPU.

**Reply by davisking** [reply]
Yeah.  To do this right requires using an object that can be either a GPU or CPU hosted block of memory under the covers so you can hide the latency of the transfer.  Which is something that can't really be hidden from the user, since they have to make a point of not interlacing CPU and GPU statements or it wrecks the throughput.

Python kinda gets away with it since python is so profoundly slow everyone is in the habit of writing code where each statement does a huge amount of work.  Since you can't have loops that do lots of work in python without it taking forever.  But in C++ it's totally normal to go into loops that do a bunch of tiny matrix operations of some kind or another, and that doesn't work at all if one of them is silently talking to the GPU since the transfer latency is so huge.

**Reply by davisking** [reply]
That is to say, I don't think a C++ library can or should hide this kind of detail since it will invariably make the user's code anomalously slow.

I do think pytorch has the right kind of API for this though, with the explicit `.to()` calls clearly flagging to the user when the transfers happen.   Not hating on pytorch, they did a wonderful job with their API.

**Reply by pfeatherstone** [reply]
Maybe this is a lot of work to get right. If we simply use cuFFT + implicit host/device management (or cuFFTW which does that for you behind an FFTW API) then the performance gains are minimal. You're better off using MKL with TBB or openMP enabled

**Reply by pfeatherstone** [reply]
Or extend `dlib::matrix` to support CUDA. All the BLAS stuff can be dealt with by cuBLAS. That seems like a ton of work.

**Reply by davisking** [reply]
Yeah it is a major deal to make `dlib::matrix` have a cuda backend. Realistically it can't be done without making the API really non-cohesive since there would invariably be a ton of things that only work for CPU matrices or CUDA matrices.  And that would only be known a runtime so there would be a lot of things that could compile but then have to emit some kind of "oh this isn't supported".

It would be better to have some other type for a gpu or cpu memory backed array and then have functions specifically setup for that that always work on both cpu and gpu.  I mean that's what the tensor class is when you get down to it.

I kinda don't like the dlib tensor class.  I've since designed much nicer stuff at work and now when I see the one in dlib it's like eh, could be better.  The second or third time you make something it's always way better.

**Comment by pfeatherstone** [comment]
> I kinda don't like the dlib tensor class. I've since designed much nicer stuff at work and now when I see the one in dlib it's like eh, could be better. The second or third time you make something it's always way better.

Do you plan on adding this to dlib? That would be awesome if we ever planned to revamp the DNN stuff into something maybe a bit more pytorch-like.

**Reply by davisking** [reply]
Na, wrote it at work so it's a bother to get it into open source.

**Reply by pfeatherstone** [reply]
Write it a 4th time and open-source it :)

**Reply by davisking** [reply]
Yeah at some point :)
